pmlc/dialect/stdx/transforms/subgroup_block_ops.cc

```diff
@@ -0,0 +1,176 @@
+    // subgroup_id, we know at this point that last index comes from the
+    // AddI operation, where second operand is subgroup_id
+    OpBuilder builder(op);
+    SmallVector<Value, 4> newIndices = op.indices();
+    auto addOp = newIndices.back().getDefiningOp();
+    newIndices.pop_back();
+    newIndices.push_back(addOp->getOperand(0));
+    auto newBlockReadOp = builder.create<SubgroupBlockReadINTELOp>(
+        op.getLoc(), op.memref(), newIndices);
+    op.replaceAllUsesWith(newBlockReadOp.getResult());
+    op.erase();
+  }
+
+  void replaceStoreOp(StoreOp op) {
+    // Replace with SubgroupBlockWriteINTELOp, update indice to remove
+    // subgroup_id, we know at this point that last indice comes from the
+    // addi operation, where second operand is subgroup_id
+    OpBuilder builder(op);
+    SmallVector<Value, 4> newIndices = op.indices();
+    auto addOp = newIndices.back().getDefiningOp();
+    newIndices.pop_back();
+    newIndices.push_back(addOp->getOperand(0));
+    builder.create<SubgroupBlockWriteINTELOp>(op.getLoc(), op.value(),
+                                              op.memref(), newIndices);
+    op.erase();
+  }
+
+  LogicalResult blockOps() {
+    mlir::Block *body = loop.getBody();
+
+    for (auto &op : body->getOperations()) {
+      // Check if the load/store ops can be replaced with subgroup block ops
+      if (auto loadOp = dyn_cast<LoadOp>(op))
+        checkLoadStore(loadOp);
+      else if (auto storeOp = dyn_cast<StoreOp>(op))
+        checkLoadStore(storeOp);
+    }
+
+    if (toReplace.empty()) {
+      // No valid vector ops used, nothing to do
+      return success();
+    }
+
+    // Do the Ops transform
+    for (auto &op : llvm::make_early_inc_range(body->getOperations())) {
+      if (auto loadOp = dyn_cast<LoadOp>(&op))
+        replaceLoadOp(loadOp);
+      else if (auto storeOp = dyn_cast<StoreOp>(&op))
+        replaceStoreOp(storeOp);
+    }
+    return success();
+  }
+};
+
+/// Returns true if no other scf.parallel ops are nested within.
+static bool isInnermostParallelOp(scf::ParallelOp parallelOp) {
+  // Only for the innermost scf.parallel op's.
+  bool isInnermost = true;
+  parallelOp.walk([&](scf::ParallelOp thisparallelOp) {
+    // Since this is a post order walk, we are able to conclude here.
+    isInnermost = (thisparallelOp == parallelOp);
+    return WalkResult::interrupt();
+  });
+  return isInnermost;
+}
+
+/// Gathers loops that have no scf.parallel nested within.
+static void gatherInnermostLoops(FuncOp f,
+                                 SmallVectorImpl<scf::ParallelOp> &loops) {
+  f.walk([&](scf::ParallelOp parallelOp) {
+    if (isInnermostParallelOp(parallelOp))
+      loops.push_back(parallelOp);
+  });
+}
+
+struct SubgroupBlockOpsPass
+    : public SubgroupBlockOpsBase<SubgroupBlockOpsPass> {
+  void runOnFunction() final {
+    auto func = getFunction();
+    SmallVector<scf::ParallelOp, 4> loops;
+    gatherInnermostLoops(func, loops);
```

Comment:
Is there is specific reason that loads/stores not in the innermost loops would be illegible for this transformation?  It seems more correct to find all loads/stores and then walk up until we hit the nearest enclosing scf.parallel (ignoring any loads/stores which are not in an scf.parallel.

Reply:
So the motivation here was to change only the ops that were changed in the subgroups devectorization pass. I can generalize this in next PR, makes sense to do it for all loads/stores as soon as we determine good way of detection of those.